lib/stdlib/src/peer.erl

```diff
@@ -0,0 +1,906 @@
+    {noreply, State#peer_state{peer_state = {down, Reason}}};
+%% default: ignore Reason and shut down normally
+maybe_stop(Reason, State) ->
+    {stop, normal, State#peer_state{peer_state = {down, Reason}}}.
+
+%% i/o protocol from origin:
+%%  * {io_reply, ...}
+%%  * {message, To, Content}
+%%  * {call, From, M, F, A}
+%%
+%% i/o port protocol, from peer:
+%%  * {io_request, From, ReplyAs, Request}
+%%  * {message, To, Content}
+%%  * {reply, From, ok | throw | error | exit | crash, Result | {Reason, Stack}}
+
+%% Handles bytes coming from alternative connection, forwarding as needed.
+handle_alternative_data(Kind, {io_request, From, FromRef, IoReq}, #peer_state{connection = Conn} = State) ->
+    %% TODO: make i/o completely async
+    Reply = {io_reply, From, FromRef, forward_request(IoReq)},
+    origin_to_peer(Kind, Conn, Reply),
+    State;
+handle_alternative_data(_Kind, {message, To, Content}, State) ->
+    To ! Content,
+    State;
+handle_alternative_data(_Kind, {reply, Seq, Class, Result}, #peer_state{outstanding = Out} = State) ->
+    {From, NewOut} = maps:take(Seq, Out),
+    gen:reply(From, {Class, Result}),
+    State#peer_state{outstanding = NewOut};
+handle_alternative_data(_Kind, {started, NodeName}, State)->
+    boot_complete(NodeName, started, State).
+
+forward_request(Req) ->
+    GL = group_leader(),
+    MRef = erlang:monitor(process, GL),
+    GL ! {io_request,self(), MRef, Req},
+    receive
+        {io_reply, MRef, Reply} ->
+            erlang:demonitor(MRef, [flush]),
+            Reply;
+        {'DOWN', MRef, _, _, _} ->
+            {error, terminated}
+    end.
+
+%% generic primitive to send data from origin to peer via alternative connection
+origin_to_peer(tcp, Sock, Term) ->
+    ok = gen_tcp:send(Sock, term_to_binary(Term));
+origin_to_peer(port, Port, Term) ->
+    true = erlang:port_command(Port, encode_port_data(term_to_binary(Term))).
+
+%% generic primitive to send data from peer to origin
+peer_to_origin(tcp, Sock, Term) ->
+    ok = gen_tcp:send(Sock, term_to_binary(Term));
+peer_to_origin(port, Port, Term) ->
+    %% converts Erlang term to terminal codes
+    %% Every binary byte is converted into two 4-bit sequences.
+    Bytes = term_to_binary(Term),
+    true = erlang:port_command(Port, encode_port_data(Bytes)).
+
+%% convert/escape Erlang term into terminal codes
+encode_port_data(Bytes) ->
+    Size = byte_size(Bytes),
+    Crc = erlang:crc32(Bytes),
+    Total = <<Size:32, Bytes/binary, Crc:32>>,
+    <<<<3:2, Upper:4, 3:2, 3:2, Lower:4, 3:2>> || <<Upper:4, Lower:4>> <= Total>>.
+
+%% convert terminal codes to Erlang term, printing everything that
+%%  was detected as text
+decode_port_data(<<>>, Str, Bin) ->
+    {Str, Bin};
+decode_port_data(<<3:2, Quad:4, 3:2, Rest/binary>>, Str, Bin) ->
+    decode_port_data(Rest, Str, quad_concat(Bin, Quad, erlang:bit_size(Bin)));
+decode_port_data(<<Char:8, Rest/binary>>, Str, Bin) ->
```

Comment:
When will this happen?

I think that this protocol allowing both binary data and text to be sent deserves some comments explaining what is going on.